src/midi/midi_alsa.cpp

```diff
@@ -0,0 +1,227 @@
+ */
+
+#include "midi_alsa.h"
+
+#if C_ALSA
+
+#include <string>
+#include <sstream>
+
+#include "string_utils.h"
+
+#define ADDR_DELIM ".:"
+
+#if (SND_LIB_MAJOR >= 1) || ((SND_LIB_MAJOR == 0) && (SND_LIB_MINOR >= 6))
+#define snd_seq_flush_output(x)           snd_seq_drain_output(x)
+#define snd_seq_set_client_group(x, name) /* nop */
+#define my_snd_seq_open(seqp)             snd_seq_open(seqp, "hw", SND_SEQ_OPEN_OUTPUT, 0)
+#else
+/* SND_SEQ_OPEN_OUT causes oops on early version of ALSA */
+#define my_snd_seq_open(seqp) snd_seq_open(seqp, SND_SEQ_OPEN)
+#endif
+
+void MidiHandler_alsa::send_event(int do_flush)
+{
+	snd_seq_ev_set_direct(&ev);
+	snd_seq_ev_set_source(&ev, my_port);
+	snd_seq_ev_set_dest(&ev, seq_client, seq_port);
+
+	snd_seq_event_output(seq_handle, &ev);
+	if (do_flush)
+		snd_seq_flush_output(seq_handle);
+}
+
+bool MidiHandler_alsa::parse_addr(const char *arg, int *client, int *port)
+{
+	std::string in(arg);
+	if (in.empty())
+		return false;
+
+	if (in[0] == 's' || in[0] == 'S') {
+		*client = SND_SEQ_ADDRESS_SUBSCRIBERS;
+		*port = 0;
+		return true;
+	}
+
+	if (in.find_first_of(ADDR_DELIM) == std::string::npos)
+		return false;
+
+	std::istringstream inp(in);
+	int val1;
+	int val2;
+	char c;
+	if (!(inp >> val1))
+		return false;
+	if (!(inp >> c))
+		return false;
+	if (!(inp >> val2))
+		return false;
+	*client = val1;
+	*port = val2;
+	return true;
+}
+
+MidiHandler_alsa::MidiHandler_alsa()
+        : MidiHandler(),
+          ev{},
+          seq_handle(nullptr),
+          seq_client(0),
+          seq_port(0),
+          my_client(0),
+          my_port(0)
+{}
+
+void MidiHandler_alsa::PlaySysex(uint8_t *sysex, size_t len)
+{
+	snd_seq_ev_set_sysex(&ev, len, sysex);
+	send_event(1);
+}
+
+void MidiHandler_alsa::PlayMsg(const uint8_t *msg)
+{
+	ev.type = SND_SEQ_EVENT_OSS;
+	ev.data.raw32.d[0] = msg[0];
+	ev.data.raw32.d[1] = msg[1];
+	ev.data.raw32.d[2] = msg[2];
+
+	unsigned char chanID = msg[0] & 0x0F;
+	switch (msg[0] & 0xF0) {
+	case 0x80:
+		snd_seq_ev_set_noteoff(&ev, chanID, msg[1], msg[2]);
+		send_event(1);
+		break;
+	case 0x90:
+		snd_seq_ev_set_noteon(&ev, chanID, msg[1], msg[2]);
+		send_event(1);
+		break;
+	case 0xA0:
+		snd_seq_ev_set_keypress(&ev, chanID, msg[1], msg[2]);
+		send_event(1);
+		break;
+	case 0xB0:
+		snd_seq_ev_set_controller(&ev, chanID, msg[1], msg[2]);
+		send_event(1);
+		break;
+	case 0xC0:
+		snd_seq_ev_set_pgmchange(&ev, chanID, msg[1]);
+		send_event(0);
+		break;
+	case 0xD0:
+		snd_seq_ev_set_chanpress(&ev, chanID, msg[1]);
+		send_event(0);
+		break;
+	case 0xE0: {
+		long theBend = ((long)msg[1] + (long)(msg[2] << 7)) - 0x2000;
+		snd_seq_ev_set_pitchbend(&ev, chanID, theBend);
+		send_event(1);
+		break;
+	}
+	default:
+		// Maybe filter out FC as it leads for at least one user to
+		// crash, but the entire midi stream has not yet been checked.
+		LOG(LOG_MISC, LOG_WARN)("ALSA: Unknown Command: %02X %02X %02X",
+		                        msg[0], msg[1], msg[2]);
```

Comment:
(I know the goal of this PR isn't to touch the code..)
`msg[0]` in this LOG should probably be `msg[0] & 0xF0`, as that's the exact command that wasn't handled.

Reply:
Hmm, I'm not sure - this log is shown in the debugger, so maybe it makes sense to show full byte. Leaving it as it is for now.